### PR TITLE
[katran]: rework inline decapsulation

### PR DESCRIPTION
**Pull request by tehnerd**
Preparation work to make GUE decap, well, works.

problem statement:
currently the way IPIP decap works is
- we process l3 header of the packet. if it is ipip - we decapsulate it
and, if dst is equal to "inline_decap" we rerun "process l3 headers"

the issue w/ gue is that we would be able to understand that this is gue
only afte processing L3 + L4 headers. and if dst is "inline_decap" we
would need to rerun parsing of L3 + L4 again.

it is very likely that running L3+L4 parsing twice would bloat bpf's
code. also for transition period (e.g. if someone is already running
inline decap, for, say, anycast service) we would need to be able to
support both IPIP and GUE decap at the same time.

Proposed solution:
recirculation. it is know technics (in networking hardware, which
needs to workaround more or less the same limitations) where, after some modificiation,
packet is "recirculated" into pipeline. we are achieving this by calling
ourself (same balancer's code, with a help of prog array + registering
outself there + bpf_tail_call) after removing outer header.

What is going to change:
before this diff we knew, if this packet was decapsulated or not, and we
did not run certain code path if it was (e.g. we did not run "source
lookup"). we did this to prevent looping of packets betwen two remote
destinations which points to each other. however we saw some corner
cases, where even w/ such protection in place, we still saw loops.
to mitigate this corner case decrement of TTL was introduced.

w/ recirculation we would have no idea if the packet was previously
decapsulated or not. that why we would start to rely on TTL decrement
all the time to fight the loops.

Tested-By:
fixed tixtures in optional tests. changed lines:

```
22:30:43.234517 IP6 (hlim 64, next-header IPIP (4) payload length: 43) 100::64 > fc00:1404::1: IP (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto UDP (17), length 43)
    192.168.1.3.31337 > 10.200.1.1.80: [udp sum ok] UDP, length 15

see decremented TTL in inner packet
22:30:43.234521 IP6 (hlim 64, next-header IPIP (4) payload length: 43) 100::bac1:103 > fc00::2307:1: IP (tos 0x0, ttl 63, id 1, offset 0, flags [none], proto UDP (17), length 43)
    192.168.1.3.31337 > 10.200.1.1.80: [udp sum ok] UDP, length 15

22:30:43.234563 IP6 (hlim 64, next-header IPv6 (41) payload length: 75) 100::64 > fc00:1404::1: IP6 (hlim 64, next-header TCP (6) payload length: 35) fc00:2307:1::2.31337 > fc00:1::1.80: Flags [.], cksum 0xda48 (correct), seq 0:15, ack 0, win 8192, length 15: HTTP

same as above. ttl is 63
22:30:43.234573 IP6 (hlim 64, next-header IPv6 (41) payload length: 75) 100::7a69:2 > fc00::2307:3: IP6 (hlim 63, next-header TCP (6) payload length: 35) fc00:2307:1::2.31337 > fc00:1::1.80: Flags [.], cksum 0xda48 (correct), seq 0:15, ack 1, win 8192, length 15: HTTP
```

### CODE REVIEW

**Comment by tehnerd** [comment]
force pushed changes: 
- adjusted counters in optional tests (as we stoped to distinct if packet was already decapsulated or not)

File: katran/lib/bpf/balancer_kern.c [hunk 1]
```diff
@@ -220,18 +213,7 @@ static inline int process_l3_headers(struct packet_description *pckt,
       // contains ip options, and we dont support em
       return XDP_DROP;
     }
-
```
Comment (udippant) [inline katran/lib/bpf/balancer_kern.c]: If have F_SRC_ROUTING but not INLINE_DECAP, then we do not decrement ttl and potentially introduce infinite forwarding of packets between two hops

File: katran/lib/bpf/balancer_kern.c [hunk 2]
```diff
@@ -346,27 +314,18 @@ static inline int process_packet(void *data, __u64 off, void *data_end,
...
       }
       data_stats->v1 += 1;
-      pckt.flags |= F_INLINE_DECAP;
-    } else {
-      // it's ipip encapsulated packet but not to decap dst. so just pass
-      // decapsulated packet to the kernel
-      return XDP_PASS;
     }
+
+    return process_encaped_pckt(&data, &data_end, xdp, &is_ipv6, &protocol);
```
Comment (udippant) [inline katran/lib/bpf/balancer_kern.c]: With INLINE_DECAP + Gue - we'll keep on looping till ttl gets to 0?

Comment (tehnerd) [reply katran/lib/bpf/balancer_kern.c]: line 308. for GUE this code path wont be triggered. as it only works for ipip/ip6ip6.

File: katran/lib/bpf/balancer_kern.c (hunk 1 above)
Comment (tehnerd) [reply katran/lib/bpf/balancer_kern.c]: also currently it is not possible to have LPM_SRC_LOOKUP (w/o this F_SRC_ROUTING is nop) w/o INLINE_DECAP: https://github.com/facebookincubator/katran/blob/master/katran/lib/bpf/balancer_consts.h#L246

**Comment by tehnerd** [comment]
force pushed a fix for usecase when backend and balancer are located on the same host.
the issue was that when we had a scenario:
"balancer send traffic to real, which is running code w/ INLINE_DECAP enabled. that code were removing outer header and than rerun balancer's code, therfore bouncing traffic back to the network (till ttl is expired)" fixing this by "reruning balancer code only if dst of the encapsulated packet is DECAP_VIP. in all other cases instead of recirculation -> pass to kernel after decap"

**Pull request closed without merging**